src/hotspot/share/runtime/vmThread.cpp

```diff
@@ -381,118 +291,188 @@ class HandshakeALotClosure : public HandshakeClosure {
   }
 };
 
-VM_Operation* VMThread::no_op_safepoint() {
-  // Check for handshakes first since we may need to return a VMop.
-  if (HandshakeALot) {
-    HandshakeALotClosure hal_cl;
-    Handshake::execute(&hal_cl);
+bool VMThread::handshake_alot() {
+  assert(_cur_vm_operation == NULL, "Already have an op");
+  assert(_next_vm_operation == NULL, "Already have an op");
+  if (!HandshakeALot) {
+    return false;
+  }
+  static jlong last_halot_ms = 0;
+  jlong now_ms = nanos_to_millis(os::javaTimeNanos());
+  // If only HandshakeALot is set, but GuaranteedSafepointInterval is 0,
+  // we emit a handshake if it's been more than a second since last.
+  jlong deadline_ms = GuaranteedSafepointInterval != 0 ? GuaranteedSafepointInterval : 1000;
+  deadline_ms += last_halot_ms;
```

Comment:
It would seem more straight-forward to write it like this, to highlight the "base" `last_halot_ms` and the "addition" `interval`:

```
 // ...
 jlong interval = (GuaranteedSafepointInterval != 0 ? GuaranteedSafepointInterval : 1000);
 jlong deadline_ms = last_halot_ms + interval;
 if (now_ms > deadline_ms) {
    last_halot_ms = now_ms;
    return true;
 }
 return false;
```

...but that's just a personal preference.

Reply:
Fixing